### PR TITLE
dev-1162 / Fix OCR matching function

### PR DESCRIPTION
**Pull request by niklub**

### CODE REVIEW

File: evalme/image/object_detection.py [hunk 1]
```diff
@@ -441,60 +441,72 @@ def distance(self, pred, local_range=1, label_weights=None, per_label=False):
 
 
 class OCREvalItem(ObjectDetectionEvalItem):
+    OCR_geo_tags = ['rectangle', 'rectangleLabels', 'brushLabels', 'polygonLabels']
```
Comment (niklub) [inline evalme/image/object_detection.py:444]: OCR_SHAPES

File: evalme/image/object_detection.py [hunk 2]
```diff
@@ -441,60 +441,72 @@ def distance(self, pred, local_range=1, label_weights=None, per_label=False):
...
             for id_pred in pred_ids:
+                # get prediction results and types
                 pred_results = pred._get_results_by_id(id_pred)
                 pred_types = self._get_types_from_results(pred_results)
-                if 'rectangle' in pred_types and 'rectangle' in gt_types:
-                    gt_results_rectangle = [item for item in gt_results if item['type'] == 'rectangle']
-                    pred_results_rectangle = [item for item in pred_results if item['type'] == 'rectangle']
-                    score = self._get_max_iou_rectangles(gt_results_rectangle, pred_results_rectangle, threshold)
+                # Tag for region selection [check OCR_geo_tags list]
+                rec_type = None
```
Comment (niklub) [inline evalme/image/object_detection.py:467]: make rec_type in XXX_types in one cycle

File: evalme/image/object_detection.py [hunk 3]
```diff
@@ -441,60 +441,72 @@ def distance(self, pred, local_range=1, label_weights=None, per_label=False):
...
+                    score = self._get_max_iou_rectangles(gt_results_rectangle, pred_results_rectangle)
                     if score < threshold:
-                        results[id_gt] = 0
+                        if not per_label:
+                            results[id_gt] = 0
                     else:
                         gt_results_labels = [item['value']['labels'] for item in gt_results if item['type'] == 'labels']
                         pred_results_labels = [item['value']['labels'] for item in pred_results if item['type'] == 'labels']
+                        # compare labels
                         if gt_results_labels == pred_results_labels:
```
Comment (niklub) [inline evalme/image/object_detection.py:484]: select the same labels before IOU calculations

File: evalme/image/object_detection.py [hunk 4]
```diff
@@ -441,60 +441,72 @@ def distance(self, pred, local_range=1, label_weights=None, per_label=False):
...
                             res = gt_results_text.match(item=pred_results_text, algorithm=algorithm)
                             if per_label:
                                 for item in pred_results_labels:
                                     for subitem in item:
-                                        results[subitem] += res
+                                        results[subitem] += res * score
                                         num_results[subitem] += 1
                             else:
-                                results[id_gt] = res
+                                results[id_gt] = res * score
```
Comment (niklub) [inline evalme/image/object_detection.py:495]: when threshold is applied, return only text distance

File: evalme/tests/test_object_detection.py [hunk 5]
```diff
@@ -471,8 +471,25 @@ def test_OCR_matching_function_not_matching_text():
...
+    ann2 = {"result": [{'id': 'buquXLcKOL', 'type': 'rectangle', 'value': {'x': 63.6, 'y': 60.92362344582593, 'width': 20.666666666666668, 'height': 10.8348134991119, 'rotation': 0}, 'origin': 'manual', 'to_name': 'image', 'from_name': 'bbox', 'image_rotation': 0, 'original_width': 768, 'original_height': 576}, {'id': 'buquXLcKOL', 'type': 'labels', 'value': {'x': 63.6, 'y': 60.92362344582593, 'width': 20.666666666666668, 'height': 10.8348134991119, 'labels': ['Text'], 'rotation': 0}, 'origin': 'manual', 'to_name': 'image', 'from_name': 'label', 'image_rotation': 0, 'original_width': 768, 'original_height': 576}, {'id': 'buquXLcKOL', 'type': 'textarea', 'value': {'x': 63.6, 'y': 60.92362344582593, 'text': ['17-RX-RR'], 'width': 20.666666666666668, 'height': 10.8348134991119, 'rotation': 0}, 'origin': 'manual', 'to_name': 'image', 'from_name': 'transcription', 'image_rotation': 0, 'original_width': 768, 'original_height': 576}]}
+    score = Metrics.apply({}, ann1, ann2, metric_name='OCR')
+    assert score > 0
```
Comment (niklub) [inline evalme/tests/test_object_detection.py:495]: explicitly defined score

Comment (KonstantinKorotaev) [reply evalme/tests/test_object_detection.py]: Fixed

File: evalme/image/object_detection.py (hunk 4 above)
Comment (KonstantinKorotaev) [reply evalme/image/object_detection.py]: Fixed

File: evalme/image/object_detection.py (hunk 3 above)
Comment (KonstantinKorotaev) [reply evalme/image/object_detection.py]: Fixed

File: evalme/image/object_detection.py (hunk 2 above)
Comment (KonstantinKorotaev) [reply evalme/image/object_detection.py]: Fixed

File: evalme/image/object_detection.py (hunk 1 above)
Comment (KonstantinKorotaev) [reply evalme/image/object_detection.py]: Fixed

File: evalme/image/object_detection.py [hunk 6]
```diff
@@ -441,60 +441,63 @@ def distance(self, pred, local_range=1, label_weights=None, per_label=False):
 
 
 class OCREvalItem(ObjectDetectionEvalItem):
+    OCR_SHAPES = ['rectangle', 'rectangleLabels', 'brushLabels', 'polygonLabels']
```
Comment (niklub) [inline evalme/image/object_detection.py:444]: check letter case vs match here

File: evalme/image/object_detection.py [hunk 7]
```diff
@@ -517,23 +520,41 @@ def _get_results_by_id(self, id):
...
         """
         res = set()
-        for result in self._raw_data:
+        for result in self._raw_data['result']:
             id = result.get('id')
             if id:
                 res.add(id)
         return list(res)
 
+    def _compare_text_tags(self, pred_types, gt_results, pred_results):
```
Comment (niklub) [inline evalme/image/object_detection.py:540]: add docstring

File: evalme/image/object_detection.py [hunk 8]
```diff
@@ -517,23 +520,41 @@ def _get_results_by_id(self, id):
         """
         Get results by ID
         """
-        res = []
-        for result in self._raw_data:
+        res = defaultdict(list)
+        for result in self._raw_data['result']:
             if result.get('id') == id:
-                res.append(result)
+                res[result.get('type')].append(result)
```
Comment (niklub) [inline evalme/image/object_detection.py:526]: result['type']

File: evalme/image/object_detection.py (hunk 6 above)
Comment (KonstantinKorotaev) [reply evalme/image/object_detection.py]: Fixed

File: evalme/image/object_detection.py (hunk 8 above)
Comment (KonstantinKorotaev) [reply evalme/image/object_detection.py]: Fixed

File: evalme/image/object_detection.py [hunk 9]
```diff
@@ -441,70 +441,89 @@ def distance(self, pred, local_range=1, label_weights=None, per_label=False):
...
+                # Tag for region selection [check OCR_SHAPES list]
+                for rec_type in OCREvalItem.OCR_SHAPES:
+                    # check if both groups have region selection tags
+                    if rec_type in pred_types and rec_type in gt_types:
+                        # check labels and compare labels
+                        gt_results_labels = gt_results.get('labels', [])
+                        pred_results_labels = pred_results.get('labels', [])
+                        if len(gt_results_labels) > 1 \
+                                or len(pred_results_labels) > 1 \
+                                or gt_results_labels[0]['value']['labels'] != pred_results_labels[0]['value']['labels']:
```
Comment (makseq) [inline evalme/image/object_detection.py:484]: What if we have more control tags than one in the labeling config? E.g. audio + ocr.

Comment (KonstantinKorotaev) [reply evalme/image/object_detection.py:484]: Works good, added test for this case

**Pull request merged**